### DISCUSSION TITLE
How do you use PSRule for Azure?

### DISCUSSION THREAD

**Discussion by BernieWhite** [Azure/PSRule.Rules.Azure#598] [Show and tell]
## 👋 Welcome!

We'd like to understand how you are using PSRule for Azure so we can dedicate time to improve it further.

Here's some suggested information to include:

- How does PSRule for Azure save you time?
- Strengths of using PSRule for Azure?
- Challenges of using PSRule for Azure?
- Have you built any custom rules for Azure?

Remember this is a public space so please don't share details of your project that are confidential.

**Comment by ianbrook** [comment]
So PSRule looks like the real thing - love the work. 
I have a couple of questions;
 - can the PSRule publish an XML to Devops Pipeline to be picked up by a 'task: PublishTestResults@2' tasks?
 - we would like to target a built artifact (compiled ARM templates from Bicep) in a pipeline, is that possible?

**Reply by ianbrook** [reply]
https://microsoft.github.io/PSRule/scenarios/validation-pipeline/validation-pipeline.html -  found this for test publishing

**Reply by BernieWhite** [reply]
@ianbrook That example is a little generic and is currently being refreshed. For Azure Pipelines you can use the [native task](https://marketplace.visualstudio.com/items?itemName=bewhite.ps-rule) with the `outputFormat: NUnit3` with an `outputPath:` and then publish the results using `'task: PublishTestResults@2`. Have a look here: https://github.com/microsoft/PSRule-pipelines/blob/main/docs/tasks.md#ps-rule-assert

> we would like to target a built artifact (compiled ARM templates from Bicep) in a pipeline, is that possible?

Yes, by default PSRule for Azure has the following behaviours:

- It won't expand `.bicep` files by default, so you can either run `bicep build` manually and analyze the resulting artifact. To do this automatically see https://aka.ms/ps-rule-azure/bicep.
- If the using `inputType: repository` or `Assert-PSRule -Format File` path specs filtered by `.gitignore` are also ignored by PSRule. This can be excluded by adding the path to `Input.PathIgnore` option with a `!` same as including a file from `.gitignore` i.e. `!out/exported/`. Alternatively build the template to a path that is not excluded.

**Reply by ianbrook** [reply]
@BernieWhite - thanks for the reply, test publishing is now working like a charm... In my research I missed finding the reference to the ps-rule-assert task (https://github.com/microsoft/PSRule-pipelines/blob/main/docs/tasks.md#ps-rule-assert) - it will help greatly
